Don't allow invalid project files!

zonkmachine opened the pull request:
The renderer hangs bad if you give it something that isn't an LMMS project as argument.
```

Loading project...
Done
|-                                 |      0%   \   <---infinite hang here!
```


This fix tests that the given file is really a file ( not a directory ). It also tests for a valid LMMS project suffix.

```
~/builds/lmms/build$ ./lmms -r ../build
The file ../build does not exist!
```
You can test specifically for if the argument is a directory and report accordingly but that would be over doing it a bit I think. 
```
~/builds/lmms/build$ ./lmms -r ../build/lmmscon
The file ../build/lmmscon does not exist!
```


```
~/builds/lmms/build$ ./lmms -r ../build/lmmsconfig.h 
../build/lmmsconfig.h is not an LMMS project fil
```


I finally moved the line `Engine::init( true );` to after the tests as it just felt like a better place.

tresf commented:
Looks good. :+1:   I'll leave time for comments.  One thought was the posiblity of loading a non-mmp or non-mmpz file, but `loadFile` only takes projects, so that's for another time. :+1:

zonkmachine commented:
> One thought was the posiblity of loading a non-mmp or non-mmpz file,

Yes, I was going to ask about it. 
I'm currently trying to move the whole file test to [here](https://github.com/LMMS/lmms/blob/master/src/core/main.cpp#L538). Right after `fileToLoad` is assigned a file.
We're currently treating file arguments for render and load differently and this should fix it. It works but I haven't tested it thoroughly. It may probably make the larger part of #2524 redundant

zonkmachine commented:
In the last version it still tests for an empty `fileToLoad` if you choose `--import` and halts with a
```The file  does not exist!```

zonkmachine commented:
![importsuffixes](https://cloud.githubusercontent.com/assets/6368949/12672778/03652c80-c679-11e5-91c3-b33dc6f7c670.png)

Same checks for files to import.

zonkmachine commented:
> One thought was the posiblity of loading a non-mmp or non-mmpz file

An mpt file?

softrabbit commented:
> An mpt file?

Any valid LMMS project file that happens to be named differently from what's expected? If I explicitly tell LMMS to load a file called "myproject.mmpx" or whatever, I expect it to at least try and not reject the file until it's at least opened the file and tried parsing it.

zonkmachine commented:
> Any valid LMMS project file that happens to be named differently from what's expected? If I explicitly tell LMMS to load a file called "myproject.mmpx" or whatever, I expect it to at least try and not reject the file until it's at least opened the file and tried parsing it.

OK. Right now if you try and save a project file with another suffix than .mmp or .mmpz, it will stick a new suffix after the one you opted for. In your case with myproject.mmpx you will get myproject.mmpx.mmpz so on my Ubuntu/Mint I have to go back and change suffix manually and then myproject.mmpx will open as you expect it. Either as a command line argument or from the gui.

tresf commented on src/core/main.cpp:
```diff
@@ -528,6 +528,57 @@ int main( int argc, char * * argv )
+					fileInfoImport.suffix().toLower() == "midi"	||
+					fileInfoImport.suffix().toLower() == "rmi"	||
+					fileInfoImport.suffix().toLower() == "flp"	||
+					fileInfoImport.suffix().toLower() == "h2song" ) )
```
Hmm.... Every time I see a laundry list of file extensions I get worried that we're making more work for ourselves down the road.  😄  Perhaps related to this PR, perhaps not.... can we start managing our extension in a centralized place?  We should be able to use `#ifdef`'s to enable/disable them globally based on what libraries/plugins cmake finds.  We already do this in several other places already.  Food for thought.

zonkmachine replied:
> Perhaps related to this PR, perhaps not.... can we start managing our extension in a centralized place?

Related but it doesn't sound like something I can implement.

tresf commented on src/core/main.cpp:
```diff
@@ -528,6 +528,57 @@ int main( int argc, char * * argv )
+	// Test file argument before continuing
+	QFileInfo fileInfoLoad( fileToLoad );
+	QFileInfo fileInfoImport( fileToImport );
+	if( !fileToLoad.isEmpty() )
```
Shouldn't this be the other way around?  Shouldn't we only instantiate `fileInfoLoad` if the QString isn't null or empty?  This would also remove the need for setting an empty QString above.

tresf replied:
The above code comment from April still remains unanswered and you have a commented line that should be removed prior to merging.

zonkmachine replied:
> Shouldn't we only instantiate fileInfoLoad if the QString isn't null or empty? This would also remove the need for setting an empty QString above.

No, I don't think so, but this part is essentially the same as before. I just moved it and expanded on the functionality.


> you have a commented line that should be removed

comment removed

jasp00 replied:
`fileInfoLoad` should be instantiated in the `if( !fileToLoad.isEmpty() )` block; `fileInfoImport`, in the `if( !fileToImport.isEmpty() )` one.

zonkmachine replied:
OK. I updated the according to Javiers suggestions.

zonkmachine commented:
Conflicts resolved.

Umcaruje commented:
@zonkmachine is there any more work needed to be done here, or it could be tested and merged?

zonkmachine commented:
Yes, mad sorry about the slow response here. Short on time. I'll look into it as soon as I possibly can.

zonkmachine commented:
OK. Trying to finish this one now.

jasp00 commented on src/core/main.cpp:
```diff
@@ -567,6 +567,57 @@ int main( int argc, char * * argv )
+		}
+		else if( !(	fileInfoLoad.suffix().toLower() == "mmp" ||
+				fileInfoLoad.suffix().toLower() == "mmpz" ||
+				fileInfoLoad.suffix().toLower() == "mpt" ) )
```
Older projects use the `xml` extension.

zonkmachine replied:
Fixed

jasp00 commented on src/core/main.cpp:
```diff
@@ -567,6 +567,57 @@ int main( int argc, char * * argv )
+		else if( !(	fileInfoImport.suffix().toLower() == "mid"	||
+				fileInfoImport.suffix().toLower() == "midi"	||
+				fileInfoImport.suffix().toLower() == "rmi"	||
+				fileInfoImport.suffix().toLower() == "flp"	||
```
FLP is dropped.

zonkmachine replied:
Fixed. Should it also drop a message that FruityLoop support is dropped since version 1.2 ?

jasp00 replied:
Such a message is useful before dropping support, when the feature is deprecated. It is not needed now.

zonkmachine replied:
Aight!

jasp00 commented:
Current fix is incomplete. Input file could be an empty file, the empty template, a broken project or a different XML file. I do not think these new checks are necessary. The check for empty projects on export should be enough.

jasp00 commented:
> I do not think these new checks are necessary.

I mean checking the extension, if the input does not exists, etc.

zonkmachine commented:
Right, back on this...

> Input file could be an empty file,

Hm, missed that. Poking around with it right now...

> the empty template

...Will open correctly.

> a broken project

A file that is empty or a broken project will be passed to LMMS who will fail to open it and report on it.

![emptyproject](https://cloud.githubusercontent.com/assets/6368949/17522544/c258cbee-5e57-11e6-99a3-4aff6d4e99bc.png)

It will then open an empty project with the title of the file project you tried to open.

jasp00 commented:
I mean that rendering those invalid projects (empty file, empty template, etc.) with `lmms -r` causes the infinite hang.

zonkmachine commented:
Right. In the case of an empty file the latest commit proposed a fix and it's working. An empty template however, I think goes a bit beyond this PR. Shouldn't it rather be catched by the renderer?

zonkmachine commented:
When exporting from the gui lmms ran TrackContainer::isEmpty() so I added this test in main.
```

$ ./lmms --rendertracks empty.mmpz
Notice: could not set realtime priority.
VST sync support disabled in your configuration
Loading project...
The project empty.mmpz is empty, aborting!

```

jasp00 commented on src/core/main.cpp:
```diff
@@ -567,6 +567,72 @@ int main( int argc, char * * argv )
+		else if( !(	fileInfoLoad.suffix().toLower() == "mmp" ||
+				fileInfoLoad.suffix().toLower() == "mmpz" ||
+				fileInfoLoad.suffix().toLower() == "mpt" ||
+				fileInfoLoad.suffix().toLower() == "xml" ) )
```
Having one of these extensions is not required; this check should be removed. The `Engine::getSong()->isEmpty()` check takes care of invalid projects.

zonkmachine replied:
OK. I initially made these suffix tests because lmms opens in an unclean manner when you throw a file whatever at it. It will open a blank project with the title you just gave it. I think at least we should not set the project title when the project couldn't be opened.

zonkmachine replied:
> It will open a blank project with the title you just gave it. I think at least we should not set the project title when the project couldn't be opened.

That's a regression from 1.1.3 which does this just fine...

jasp00 commented on src/core/main.cpp:
```diff
@@ -567,6 +567,72 @@ int main( int argc, char * * argv )
+		else if( !(	fileInfoImport.suffix().toLower() == "mid"	||
+				fileInfoImport.suffix().toLower() == "midi"	||
+				fileInfoImport.suffix().toLower() == "rmi"	||
+				fileInfoImport.suffix().toLower() == "h2song" ) )
```
The import plug-ins know better how to handle these files; this check should be removed. `ImportFilter::import()` takes care of this.

jasp00 commented on src/core/main.cpp:
```diff
@@ -567,6 +567,72 @@ int main( int argc, char * * argv )
+	if( !fileToLoad.isEmpty() )
+	{
+		QFileInfo fileInfoLoad( fileToLoad );
+		if( !fileInfoLoad.exists() )
```
`!fileInfoLoad.size()` checks if the file does not exist or its size is zero. The error message could be `The file %s does not have any content`.

zonkmachine replied:
Fixed!

jasp00 commented on src/core/main.cpp:
```diff
@@ -567,6 +567,72 @@ int main( int argc, char * * argv )
+	else if( !fileToImport.isEmpty() )
+	{
+		QFileInfo fileInfoImport( fileToImport );
+		if( !fileInfoImport.exists() )
```
`!fileInfoImport.size()` could be done here.

zonkmachine replied:
Fixed!

jasp00 commented on src/core/main.cpp:
```diff
@@ -567,6 +567,72 @@ int main( int argc, char * * argv )
+			printf( "%s is a directory.\n",
+					fileToImport.toStdString().c_str() );
+			exit( 1 );
+		}
```
These two tests (`size()` and `isDir()`) should be in a function.

zonkmachine replied:
Aight! That cleaned things up. The PR shrunk by 50%... :)

zonkmachine commented on src/core/main.cpp:
```diff
@@ -567,6 +588,15 @@ int main( int argc, char * * argv )
+	{
+		fileCheck( fileToLoad );
+	}
+	if( !fileToImport.isEmpty() )
```
```c++
else if
```

Should I short this one out if we have one of them and give fileToLoad priority?

zonkmachine replied:
Yeah, fixed that...

zonkmachine commented on src/core/main.cpp:
```diff
@@ -636,15 +666,13 @@ int main( int argc, char * * argv )
+		{
+			printf("The project %s is empty, aborting!\n", fileToLoad.toUtf8().constData() );
+			exit( EXIT_FAILURE );
+		}
```
Is exit() even a good way to terminate the application at this point?

jasp00 replied:
You may omit `exit()` and avoid the export, so an upgraded `.lmmsrc.xml` can be written, but `exit()` (or `return`) is not a bad way.

zonkmachine replied:
oK. I just skimmed past some comment on the net that exit() could be a problem with multi threaded applications. I'm happy with .lmmsrc.xml not being written so we keep failing loads from appearing in the recently opened projects list.

zonkmachine commented:
lmms used to load default.mmp on failing to load a project but this was broken in commit: https://github.com/LMMS/lmms/commit/2c7036edd6efd91ee3394fa442252c67e437a307#diff-1c0d925f7437a8e3a837c0a991b4cf34

I'll try and look into the original issue behind that PR

zonkmachine commented:
It's better to not inflate this PR. The regression in Song.cpp is not directly related to these fixes so I'll treat them separately.

Rebased and squashed! :tractor:

jasp00 commented:
This request may be merged.

zonkmachine commented:
:tractor:
